File: baekjoon/Python/2252_줄 세우기.py
from collections import deque
N, E = map(int, input().split())
A = [[] for _ in range(N + 1)] # 비교 데이터 저장 인접 리스트
indegree = [0] * (N + 1) # 진입 차수 리스트

# 인접 리스트 데이터 저장
for i in range(E):
    S, E = map(int, input().split())
    A[S].append(E)
    indegree[E]+=1 # 진입차수 데이터 저장

queue = deque()
for i in range(1, N + 1): # 진입 차수 값 0인 노드 큐에 삽입
    if indegree[i] == 0:
        queue.append(i)
        
while queue: # 위상 정렬
    now = queue.popleft()
    print(now, end = ' ')
    # 현재노드 연결노드
    for next in A[now]:
        indegree[next] -= 1 # 연결노드 진입차수 감소
        if indegree[next] == 0:
            queue.append(next)

File: baekjoon/Java/2252_줄 세우기.java
import java.util.*;

public class Main {
    public static void main(String[] args) throws Exception {
        Scanner sc = new Scanner(System.in);
        int N = sc.nextInt();
        int E = sc.nextInt();
        ArrayList<ArrayList<Integer>> A = new ArrayList<>();
        Queue<Integer> queue = new LinkedList<>(); // 위상 정렬
        
        for (int i = 0; i <= N; i++){
            A.add(new ArrayList<>());
        }
        int[] indegree = new int[N + 1]; // 진입차수 배열
        for (int i = 0; i < E; i++){
            int S = sc.nextInt();
            int M = sc.nextInt();
            A.get(S).add(M);
            indegree[M]++; // 진입 차수 배열 데이터 저장
        }
        for (int i = 1; i <= N; i++){
            if (indegree[i] == 0) queue.offer(i);
        }
        while (!queue.isEmpty()) {
            int now = queue.poll();
            System.out.print(now + " ");
            for (int next : A.get(now)){
                indegree[next]--;
                if (indegree[next] == 0) queue.offer(next);
            }
        }
    }
}
